nengo/tests/test_params.py

```diff
@@ -325,42 +325,38 @@ class Test(object):
     assert set(params.iter_params(Test())) == {'p1', 'p2'}
 
 
-def test_all_nengo_parameters_can_be_configured():  # noqa: C901
+def test_configure_all_nengo_parameters():
+
+    # make up a non-default value for the parameter
+    conv_func = {
+        params.BoolParam: lambda attr: not attr.default,
+        params.NumberParam: lambda attr: (
+            1 if attr.default is None else attr.default + 1),
+        params.StringParam: lambda attr: "abc",
+        params.NdarrayParam: lambda attr: np.zeros([1] * len(attr.shape)),
+        nengo.base.ProcessParam: lambda attr: nengo.processes.WhiteNoise(),
+        nengo.node.OutputParam: lambda attr: lambda x: x + 1,
+        nengo.synapses.SynapseParam: lambda attr: nengo.synapses.Alpha(0.1),
+        nengo.solvers.SolverParam: lambda attr: nengo.solvers.LstsqL2nz(
+            weights=isinstance(attr, nengo.connection.ConnectionSolverParam)),
+        nengo.connection.ConnectionFunctionParam: lambda attr: lambda x: x + 1,
+        nengo.learning_rules.LearningRuleTypeParam: (
+            lambda attr: nengo.learning_rules.PES()),
+        nengo.neurons.NeuronTypeParam: lambda attr: nengo.AdaptiveLIF(),
+    }
+
     with nengo.Network() as net:
+        # NB: Probe must go before Connection
```

Comment:
What does NB stand for?
Why must the `Probe` go before connection?
What do you think about using `pytest.mark.parametrize` to eliminate the outer for loop?

Reply:
btw: I like the dictionary approach. :)

Reply:
NB basically mean [important note](https://en.wikipedia.org/wiki/Nota_bene), thought it was relatively common but I can change it if it's not.

`Probe.solver` uses connection's default by default. I had changed the for loop to use `net.objects` (which has all the types) instead of the list and got failures because the probe solver was using weights; this was because the connection default had changed, and the probe solver uses the connection's default by default.

IMO `parametrize` would inflate the number of tests by a fair bit, and doesn't feel like it'd be any more readable, so I would opt for keeping it as is.

Reply:
> Probe.solver uses connection's default by default. I had changed the for loop to use net.objects (which has all the types) instead of the list and got failures because the probe solver was using weights; this was because the connection default had changed, and the probe solver uses the connection's default by default. I added the note so that other people wouldn't try to do the same and get failures due to dict ordering.

Shouldn't that be solved by moving the `with nengo.Network() as net:` inside the for loop? That seems better to me because it isolates the indivdual tests and the order of iteration does not matter. (If I'm not overlooking something.)



> IMO parametrize would inflate the number of tests by a fair bit, and doesn't feel like it'd be any more readable, so I would opt for keeping it as is.

The output for a test failure might be more readable, but I would have to actually try that out and not sure if I'm feeling that strongly about it ...

Reply:
> Shouldn't that be solved by moving the with nengo.Network() as net: inside the for loop?

I don't think that would matter, as you're always within that network's context, so if the connection default is set before the probe default, it will use it.

You could get around it by simply not being in the network's context though; it's not actually necessary for anything the test is doing. Just tested it and it works, so I'll push that change.